File: esp32/components/graphics_lib/graphics/screen/screen.cpp
#include "screen.h"

void Screen::add_element(std::unique_ptr<UIElement> element)
{
    elements.push_back(std::move(element));
}

void Screen::on_scroll(const bool direction)
{
    if (focus)
    {
        elements[row_offset]->on_event(
            direction ? EVENT_SCROLL_RIGHT : EVENT_SCROLL_LEFT
        );
    } else
    {
        if (direction)
        {
            row_offset = std::min(row_offset + 1, elements.size() - 1);
        }
        else
        {
            if (row_offset == 0) return;
            row_offset--;
        }
    }
}

void Screen::on_click()
{
    focus = elements[row_offset]->on_event(EVENT_CLICK);
}

bool Screen::on_back()
{
    if (focus)
    {
        focus = false;
        return elements[row_offset]->on_event(EVENT_BACK);
    }
    return false;
}

bool Screen::on_custom_event(uint32_t)
{
    return false;
}

/*
const std::vector<std::unique_ptr<UIElement>>& Screen::get()
{
    return elements;
}

bool Screen::is_empty() const
{
    return elements.empty();
}

UIElement& Screen::get_element(const uint64_t offset) const
{
    return *elements.at(offset);
}

size_t Screen::size() const
{
    return elements.size();
}

bool Screen::on_event(const graphics_event_t event, const uint16_t element)
{
    return elements[element]->on_event(event);
}
*/

std::vector<std::string> Screen::render(const uint8_t rows_size)
{
    std::vector<std::string> rows_render{};

    const size_t offset = row_offset - (row_offset % 2);

    for (size_t i = offset; i < std::min(offset + rows_size, elements.size()); ++i)
    {
        rows_render.push_back(elements[i]->render(i == row_offset));
    }
    return rows_render;
}

File: esp32_rust/components/main_rust/src/graphics/screen.rs
use crate::graphics::event::GraphicsEvent;
use crate::graphics::ui::element::{UIElement, UIElementState};
use crate::graphics::ui::text::UIText;

pub struct ScreenData {
    pub elements: Vec<Box<dyn UIElement>>,
    pub row_offset: usize,
    pub focus: bool
}

impl ScreenData {
    pub fn new() -> ScreenData {
        ScreenData {
            elements: Vec::new(),
            row_offset: 0,
            focus: false
        }
    }

    pub fn add_element<T>(&mut self, item: T) where T: UIElement + 'static {
        self.elements.push(Box::new(item));
    }

    pub fn add_text(&mut self, text: String) {
        self.elements.push(Box::new(UIText::new(text)));
    }

    pub fn clear(&mut self) {
        self.elements.clear();
    }
}

pub trait Screen {
    fn get_data(&self) -> &ScreenData;
    fn get_data_mut(&mut self) -> &mut ScreenData;

    fn on_start(&mut self) {
        let data = self.get_data_mut();
        data.row_offset = 0;
        for elem in data.elements.iter_mut() {
            elem.on_event(GraphicsEvent::ScreenStart);
        }
    }
    fn on_end(&mut self) {
        for elem in self.get_data_mut().elements.iter_mut() {
            elem.on_event(GraphicsEvent::ScreenEnd);
        }
    }

    fn on_click(&mut self) {
        let data = self.get_data_mut();
        let elem = data.elements.get_mut(data.row_offset).expect("out of range");
        data.focus = elem.on_event(GraphicsEvent::Click);
    }

    fn on_back(&mut self) -> bool {
        let data = self.get_data_mut();
        if data.focus {
            data.focus = false;
            return data.elements[data.row_offset].on_event(GraphicsEvent::Back);
        }
        false
    }

    fn on_scroll(&mut self, direction: bool) {
        let data = self.get_data_mut();
        let row_offset = data.row_offset;
        if data.focus {
            let event = if direction {
                GraphicsEvent::ScrollRight
            } else {
                GraphicsEvent::ScrollLeft
            };
            data.elements[row_offset].on_event(event);
        } else {
            if direction {
                data.row_offset = (row_offset + 1).min(data.elements.len() - 1);
            } else {
                if row_offset == 0 {
                    return;
                }
                data.row_offset -= 1;
            }
        }
    }

    fn on_custom_event(&mut self, _event: u32) -> bool {
        false
    }

    fn refresh(&mut self) {}

    fn render(&self, selector_press: bool) -> Vec<String> {
        let data = self.get_data();
        let i = data.row_offset;
        let len = data.elements.iter().len();
        let end = (i + 2).min(len);

        if i < len {
            data.elements[i..end]
                .iter()
                .enumerate()
                .map(|(index, item)| item.render(
                    if index == 0 {
                        if selector_press {
                            UIElementState::SelectorPress
                        } else {
                            UIElementState::Selected
                        }
                    } else {
                        UIElementState::None
                    }
                ))
                .collect()
        } else {
            vec![]
        }
    }
}
